src/rulesets/Base/mapreduce.jl

```diff
@@ -136,3 +136,93 @@ function ∇prod_one_zero!(dx, x, dy::Number=1)  # Assumes exactly one x is zero
     dx[i_zero] += p_rest * dy
     return
 end
+
+#####
+##### `cumprod`
+#####
+
+function rrule(::typeof(cumprod), x::AbstractVector{<:Real}; dims=1)
+    y = cumprod(x; dims=dims)  # does nothing unless dims == 1
+    function cumprod_pullback_1(dy)
+        dx_thunk = InplaceableThunk(
+            @thunk if dims == 1
+                ∇cumprod(x, dy, y)
+            else
+                dy
+            end
+            ,
+            dx -> if dims == 1
+                ∇cumprod!(dx, x, dy, y)
+            else
+                dx .+= dy
+            end
+            )
+        return (NO_FIELDS, dx_thunk)
+    end
+    return y, cumprod_pullback_1
+end
+
+function rrule(::typeof(cumprod), x::AbstractArray{<:Real}; dims)
+    y = cumprod(x; dims=dims)
+    @assert dims isa Integer
```

Comment:
why not put this in the signature?

Reply:
I think I was thinking that I'd rather it fail on `cumprod` than on `rrule`, so this mostly a note to readers.